### PR TITLE
docker support

### PR DESCRIPTION
**Pull request by Paper-Dragon**
#59  I've written some of the necessary files for docker support, through some commands to run normally through. But there is still much to be desired.
- must be add `--privileged` option in docker run command
```bash
[2023-08-30T07:02:28Z INFO  tun2proxy] Proxy socks5 server: 162.0.220.216:46631
[2023-08-30T07:02:28Z INFO  tun2proxy::setup] [1] Setting up network configuration
[2023-08-30T07:02:28Z ERROR tun2proxy::setup] nix::errno::Errno EPERM
[2023-08-30T07:02:28Z INFO  tun2proxy::setup] [7] Restoring network configuration
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: OSError(EPERM)', src/setup.rs:269:29
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-08-30T07:02:28Z ERROR tun2proxy] Failed to read from pipe
```

I don't think this option is necessary. IDK, maybe is a bug.

- I also found 0x80 error in master branch source code . #53 
```bash
[2023-08-30T07:04:58Z INFO  tun2proxy] Proxy socks5 server: 162.0.220.216:46631
[2023-08-30T07:04:58Z INFO  tun2proxy::setup] [1] Setting up network configuration
[2023-08-30T07:04:58Z INFO  tun2proxy::tun2proxy] Unsupported protocol 0x80 in IP packet, ignored
[2023-08-30T07:05:02Z INFO  tun2proxy::tun2proxy] Unsupported protocol 0x80 in IP packet, ignored
[2023-08-30T07:05:10Z INFO  tun2proxy::tun2proxy] Unsupported protocol 0x80 in IP packet, ignored
[2023-08-30T07:05:25Z INFO  tun2proxy::tun2proxy] Unsupported protocol 0x80 in IP packet, ignored
```

Possible build scripts.
```bash
docker build -t tun2proxy:version .
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -102,6 +102,30 @@ Currently, tun2proxy supports HTTP, SOCKS4/SOCKS4a and SOCKS5. A proxy is suppli
 URL format. For example, an HTTP proxy at `1.2.3.4:3128` with a username of `john.doe` and a password of `secret` is
 supplied as `--proxy http://john.doe:secret@1.2.3.4:3128`. This works analogously to curl's `--proxy` argument.
 
+## Docker Support
+
+```bash
+docker run -d \
+	-e PROXY=PROXY_TYPE://PROXY_IP:PROXY_PORT \
+	-v /dev/net/tun:/dev/net/tun \
```
Comment (planetoryd) [inline README.md:110]: looks weird. does it work ?

Comment (Paper-Dragon) [reply README.md:110]: Running container if add `--privileged` option in docker run command.

Comment (Paper-Dragon) [reply README.md:110]: - 添加--privileged可以正常运行。
- 映射设备是因为内核tun模块需要。
- 当然，不映射也可以，使用 --privileged 特权在容器内创建tun设备也可以。

**Comment by Paper-Dragon** [comment]
i build a example docker image
### Example: use privileged: test passed
```bash
docker run -itd --name proxy \
    --privileged \
    -e PROXY=socks5://nscoks1:xxxxx@162.0.220.216:46631 \
    --sysctl net.ipv6.conf.all.disult.disable_ipv6=0 \
    jockerdragon/tun2proxy:test
```

```bash
docker run -it --rm \
    --network "container:proxy" \
    curlimages/curl cip.cc
IP      : 143.55.129.203
地址    : 美国  纽约州  纽约
运营商  : fitnyc.edu

数据二  : 美国 | 纽约技术大学

数据三  : 美国纽约纽约

URL     : http://www.cip.cc/143.55.129.203

````
### Example: not use privileged: much error
```bash
docker run -itd --name proxy  -e PROXY=socks5://nscoks1:Cavxxxx@162.0.220.216:46631  jockerdragon/tun2proxy:test
docker logs proxy  
[2023-09-01T04:56:51Z INFO  tun2proxy] Proxy socks5 server: 162.0.220.216:46631
[2023-09-01T04:56:51Z INFO  tun2proxy::setup] [1] Setting up network configuration
[2023-09-01T04:56:51Z ERROR tun2proxy::setup] [7] Command `ip tuntap add name tun0 mode tun` failed: open: No such file or directory

[2023-09-01T04:56:51Z INFO  tun2proxy::setup] [7] Restoring network configuration
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: OSError(EPERM)', src/setup.rs:269:29
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-09-01T04:56:51Z ERROR tun2proxy] Failed to read from pipe
```
then add ` -v /dev/net/tun:/dev/net/tun`
```bash
root@k8s-server:/home/k8s/tun2proxy# docker run -itd --name proxy  -e PROXY=socks5://nscoks1:Cavcxxxx@162.0.220.216:46631 -v /dev/net/tun:/dev/net/tun jockerdragon/tun2proxy:test
4aa513d1bdec2d0f229892f39bb21375668dd8e73786878e2628785f0b65bdcd
root@k8s-server:/home/k8s/tun2proxy# docker logs 4a
[2023-09-01T04:59:13Z INFO  tun2proxy] Proxy socks5 server: 162.0.220.216:46631
[2023-09-01T04:59:13Z INFO  tun2proxy::setup] [1] Setting up network configuration
[2023-09-01T04:59:13Z ERROR tun2proxy::setup] [6] Command `ip tuntap add name tun0 mode tun` failed: ioctl(TUNSETIFF): Operation not permitted

[2023-09-01T04:59:13Z INFO  tun2proxy::setup] [6] Restoring network configuration
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: OSError(EPERM)', src/setup.rs:269:29
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-09-01T04:59:13Z ERROR tun2proxy] Failed to read from pipe
```
I think because not had NET_ADMIN privileged. then add `--cap-add NET_ADMIN`
```bash
root@k8s-server:/home/k8s/tun2proxy# docker run -itd --name proxy  -e PROXY=socks5://nscoks1:Caxxxx@162.0.220.216:46631 -v /dev/net/tun:/dev/net/tun --cap-add NET_ADMIN jockerdragon/tun2proxy:test
c28808227267cfa3ec65ab8de73ed54760123745fa98216f910def05c994a9c6
root@k8s-server:/home/k8s/tun2proxy# docker logs c28
[2023-09-01T05:01:26Z INFO  tun2proxy] Proxy socks5 server: 162.0.220.216:46631
[2023-09-01T05:01:26Z INFO  tun2proxy::setup] [1] Setting up network configuration
[2023-09-01T05:01:26Z ERROR tun2proxy::setup] nix::errno::Errno EPERM
[2023-09-01T05:01:26Z INFO  tun2proxy::setup] [7] Restoring network configuration
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: OSError(EPERM)', src/setup.rs:269:29
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2023-09-01T05:01:27Z ERROR tun2proxy] Failed to read from pipe
```
### I don't know why it doesn't work properly in unprivileged mode

**Comment by blechschmidt** [comment]
At first I didn't understand the use case, but having tun2proxy serve as a proxy for other Docker containers is quite cool. Sorry for the late review and thanks a lot for the contribution!

**Pull request merged**